File: backend/app/models/book.rb
class Book < ApplicationRecord
  GENRES = [
    "Fiction", "Dystopian", "Science", "Philosophy",
    "Science Fiction", "History", "Technology", "Fantasy", "Biography"
  ].freeze

  has_many :borrowings, dependent: :destroy

  validates :title, :author, presence: true
  validates :genre, presence: true, inclusion: { in: GENRES, message: "%{value} is not a valid genre" }
  validates :isbn, presence: true, uniqueness: true
  validates :total_copies, numericality: { greater_than_or_equal_to: 0 }

  def available_copies
    total_copies - borrowings.active.count
  end
end

File: frontend/src/types/book.ts
export interface Book {
  id: number;
  title: string;
  author: string;
  genre: string;
  isbn: string;
  total_copies: number;
  available_copies: number;
  created_at: string;
}

export interface BookFormData {
  title: string;
  author: string;
  genre: string;
  isbn: string;
  total_copies: number;
}
